staging/src/k8s.io/apiserver/pkg/util/flowcontrol/metrics/sample-and-count.go

```diff
@@ -0,0 +1,193 @@
+}
+
+// Generate makes a new RatioedChangeObserver
+func (swg *sampleAndCountObserverGenerator) Generate(initialNumerator, initialDenominator float64, labelValues []string) RatioedChangeObserver {
+	ratio := initialNumerator / initialDenominator
+	when := swg.clock.Now()
+	countLabelValues := [][]string{}
+	for idx := range swg.countBuckets {
+		countLabelValues = append(countLabelValues, append([]string{swg.bucketLabelValues[idx]}, labelValues...))
+	}
+	return &sampleAndCountHistograms{
+		sampleAndCountObserverGenerator: swg,
+		labelValues:                     labelValues,
+		countLabelValues:                countLabelValues,
+		denominator:                     initialDenominator,
+		sampleAndCountAccumulator: sampleAndCountAccumulator{
+			lastSet:    when,
+			lastSetInt: swg.quantize(when),
+			numerator:  initialNumerator,
+			ratio:      ratio,
+			lastBucket: findBucket(swg.countBuckets, ratio),
+		}}
+}
+
+func (swg *sampleAndCountObserverGenerator) Metrics() Registerables {
+	return Registerables{swg.samples, swg.counts}
+}
+
+type sampleAndCountHistograms struct {
+	*sampleAndCountObserverGenerator
+	labelValues      []string
+	countLabelValues [][]string // one []string for each bucket
+
+	sync.Mutex
+	denominator float64
+	sampleAndCountAccumulator
+}
+
+type sampleAndCountAccumulator struct {
+	lastSet    time.Time
+	lastSetInt int64 // lastSet / samplePeriod
+	numerator  float64
+	ratio      float64 // numerator/denominator
+	lastBucket int     // negative when not in a bucket
+}
+
+var _ RatioedChangeObserver = (*sampleAndCountHistograms)(nil)
+
+func (saw *sampleAndCountHistograms) Add(deltaNumerator float64) {
+	saw.innerSet(func() {
+		saw.numerator += deltaNumerator
+	})
+}
+
+func (saw *sampleAndCountHistograms) Observe(numerator float64) {
+	saw.innerSet(func() {
+		saw.numerator = numerator
+	})
+}
+
+func (saw *sampleAndCountHistograms) SetDenominator(denominator float64) {
+	saw.innerSet(func() {
+		saw.denominator = denominator
+	})
+}
+
+func findBucket(countBuckets []float64, x float64) int {
+	if x < countBuckets[0] {
+		return -1
+	}
+	var idx int
+	for idx = len(countBuckets) - 1; idx > 0 && x < countBuckets[idx]; idx-- {
+	}
+	return idx
+}
+
+func (saw *sampleAndCountHistograms) innerSet(updateNumeratorOrDenominator func()) {
+	when, whenInt, acc, wellOrdered := func() (time.Time, int64, sampleAndCountAccumulator, bool) {
+		saw.Lock()
+		defer saw.Unlock()
+		// Moved these variables here to tiptoe around https://github.com/golang/go/issues/43570 for #97685
+		when := saw.clock.Now()
+		whenInt := saw.quantize(when)
+		acc := saw.sampleAndCountAccumulator
+		dt := when.Sub(acc.lastSet)
+		wellOrdered := dt >= 0
+		updateNumeratorOrDenominator()
+		saw.ratio = saw.numerator / saw.denominator
+		if wellOrdered {
+			bucket := findBucket(saw.countBuckets, saw.ratio)
+			if saw.lastBucket >= 0 {
+				saw.counts.WithLabelValues(saw.countLabelValues[saw.lastBucket]...).Add(dt.Seconds())
```

Comment:
This has a problem that we report it only for a given bucket.
This seems fine as long as we name the bucket in a clear way e.g. not just the end of the bucket, but rather the whole bucket.
Something like "0.9-1.0" (or sth like that).

Reply:
https://prometheus.io/docs/practices/histograms/#quantiles

Reply:
See my response above - we don't want quantiles from the observations.